### PR TITLE
WIP: Redux migration

### PR DESCRIPTION
**Pull request by KamranAsif**
## Description

Migrating scrollHelper to redux store
## Motivation and Context

https://github.com/schrodinger/fixed-data-table-2/issues/37

Progress: 
- [x] Inject redux store logic into FDT calculateState
- [ ] Migrate scrollHelper state to redux store
- [ ] Create actions/reducers for scrollHelper setters
- [ ] Inject store updating to replace getters
- [x] Cleanup BufferedRows state

### CODE REVIEW

**Comment by KamranAsif** [comment]
Got basic scrolling working. Next step is to try and remove the old scrollHelper entirely, but this requires removing FixedDataTableRowBuffer

**Comment by KamranAsif** [comment]
Managed to get rid of BufferedRows state. scrollHelper needs a major refactor, but I'm punting that until all the logic is migrated.

Not sure what kind of pattern to use, but having to pass in the state to every function is less than ideal.
 
I also end up creating a new object for every change, so no diffing benefit in yet. Need to figure that out as well.

File: src/FixedDataTable.js [hunk 1]
```diff
@@ -327,29 +327,35 @@ var FixedDataTable = React.createClass({
...
+        rows,
+        rowHeights,
         scrollContentHeight,
-        maxScrollY
+        scrollY,
       });
-    });
+    };
+    FixedDataTableStore.subscribe(update);
+    setTimeout(update);
```
Comment (KamranAsif) [inline src/FixedDataTable.js]: This is pretty hacky. I need to merge update and calculateState

**Comment by wcjordan** [comment]
Also did we ever find out if anything bad happens if multiple Redux instances exist on a page?  This [thread](https://github.com/reactjs/redux/issues/871) makes it seem like it's fine, but I wanted to confirm.

**Comment by KamranAsif** [comment]
From what I tested and researched, I couldn't find anything wrong with it

File: src/reducers/scroller.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  switch (action.type) {
+    case ActionTypes.PROP_CHANGE: 
+      let { props } = action;
+
+      if (props.rowsCount === state.rowsCount && props.rowHeight === state.rowHeight) {
+        return state;
+      }
+
+      //TODO (asif) reinitalize when viewheight changes
+      var a = scrollHelper.initailize(state, props);
```
Comment (wcjordan) [inline src/reducers/scroller.js]: better variable name than a

File: src/reducers/scrollHelper.js [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+}
+
+/**
+ * Creates new state object when rowHeight or rowCount changes
+ *
+ * @param {!Object} state
+ * @param {!Object} props
+ * @return {!Object}
+ */
+function initailize(state, props) {
```
Comment (wcjordan) [inline src/reducers/scrollHelper.js]: typo, s/b initialize

File: src/reducers/scrollHelper.js [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+    MIN_BUFFER_ROWS,
+    MAX_BUFFER_ROWS
+  );
+
+  let storedHeights = new Array(rowsCount);
+  for (var i = 0; i < rowsCount; i++) {
+    storedHeights[i] = rowHeight;
+  }
+
+  state = Object.assign({}, state, {
```
Comment (wcjordan) [inline src/reducers/scrollHelper.js]: Is there any documentation on what all these fields are and how they are / can be updated / changed?  It's currently a challenge to follow the logic here.

**Comment by wcjordan** [comment]
Sorry for the delay looking at this.  It looks good as far as I can tell, although FixedDataTable.js isn't in a state where I could review it and I could use some documentation or overview of scrollHelper to understand it better.

Also if you have any documentation / plans for next steps, that could help me see how this fits into the big picture.

**Comment by tomjnsn** [comment]
Quick question on this pull request: is this definitely planned to be pulled in once it is complete? Just asking because I'm using mobx in a project that I have fixed-data-table-2 in and would rather not carry around the weight of both redux and mobx and am concerned that there are going to be issues with mixing the two together.

**Comment by wcjordan** [comment]
While not definite, this is the most likely direction for the project since scroll state is proving difficult to manage.  It currently suffers from bugs in cases with dynamic row heights which we'd like to fix.  We do like hearing feedback though and would be interested if anyone in the community has alternative ideas.

With regards to weight of redux, adding redux will add 2.7 kB to the 75+ kB FDT.  While this is non-trivial, it seems like a reasonable trade-off.  We do want to avoid bringing in redux-react which seems overweight at an additional 3.5 kB, and we'll also avoid bringing in additional dependencies from the Redux ecosystem unless there's great need.

With regards to mixing MobX and Redux, I'm not aware of any issues which would arise from including both on a page.  From my understanding of their architectures I don't think they conflict, but it'd be great if you could try this branch out as it gets closer to stable and let us know your experience.

**Comment by tomjnsn** [comment]
Sounds like a decent tradeoff between size and what it will get you. I'll look forward to when you've got it to a point that I can test it with mobx. Thanks for keeping the fixed-data-table component moving forward!

File: src/reducers/scroller.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+const DEFAULT_STATE = {
+  firstRowIndex: 0,
+  firstRowOffset: 0,
+  scrollY: 0,
+  scrollContentHeight: 0,
+  maxScrollY: 0
+};
+
+const NO_ROWS_SCROLL_RESULT = {
```
Comment (wcjordan) [inline src/reducers/scroller.js]: unused here, duplicated from scrollhelper

**Comment by KamranAsif** [comment]
Closing this in favor of https://github.com/schrodinger/fixed-data-table-2/pull/77 since it includes these commits

**Comment by LESARQ** [comment]
Why do we need to inline specify the height? Why not just allowing users to set it with CSS/SCSS/etc?
Wouldn't it be easier if you wanted, for example, to have dynamic Cell/Row height using perhaps "flex".

I think that this way it would be more customizable per user. We could still set fixed Height's.

Because I'm struggling with dynamic row/cell heights, and I achieved what I wanted with `CellMeasurer`, however it doesn't fell right, it seems just a hack to go around a "flawed" design.

Or is there any reason that I don't understand?
I'm not 100% into React and perhaps I can't understand if there are implications that invalidates my idea.

**Comment by wcjordan** [comment]
Since the rows are virtualized the DOM doesn't have access to all the row heights.  Because of this, we have to make a choice between a scrollbar which is correctly positioned and sized or the cell heights being sized via CSS / content.

In the future we could experiment with a mode where the scrollbar position is allowed to be inaccurate, but cell heights are easier.  But for now it's not on the roadmap.

**Pull request closed without merging**